### PR TITLE
improves privileges matching by using a map (O(1) instead of O(n))

### PR DESCRIPTION
**Pull request by seut**
Privileges are now stored into a map for quick lookup by hashCode.
Decreases privilege matching time complexity from O(n) to O(1).
Also finding any privilege for a certain class results now in a simple
map lookup by ident (O(1)).

### MASTER
```
Benchmark                                                (privilegesSize)  Mode  Cnt        Score       Error  Units
UserPrivilegesBenchmark.measureTableMatches                           100  avgt   50     4051.418 ±    22.249  ns/op
UserPrivilegesBenchmark.measureTableMatches                          1000  avgt   50    47165.984 ±   651.951  ns/op
UserPrivilegesBenchmark.measureTableMatches                         10000  avgt   50   800149.762 ± 17286.495  ns/op
UserPrivilegesBenchmark.measureTableMatchesAnyOfCluster               100  avgt   50     8221.696 ±    97.207  ns/op
UserPrivilegesBenchmark.measureTableMatchesAnyOfCluster              1000  avgt   50    85308.593 ±   553.596  ns/op
UserPrivilegesBenchmark.measureTableMatchesAnyOfCluster             10000  avgt   50  1528666.135 ± 17420.871  ns/op
UserPrivilegesBenchmark.measureTableMatchesAnyOfSchema                100  avgt   50     8162.619 ±    50.064  ns/op
UserPrivilegesBenchmark.measureTableMatchesAnyOfSchema               1000  avgt   50    89393.748 ±   730.441  ns/op
UserPrivilegesBenchmark.measureTableMatchesAnyOfSchema              10000  avgt   50  1518160.490 ± 27210.074  ns/op
UserPrivilegesBenchmark.measureTableMatchesAnyOfTable                 100  avgt   50     3941.507 ±    37.448  ns/op
UserPrivilegesBenchmark.measureTableMatchesAnyOfTable                1000  avgt   50    45257.212 ±   568.994  ns/op
UserPrivilegesBenchmark.measureTableMatchesAnyOfTable               10000  avgt   50   797699.735 ± 12184.941  ns/op
UserPrivilegesBenchmark.measureTableMatchesCluster                    100  avgt   50     8270.019 ±   138.425  ns/op
UserPrivilegesBenchmark.measureTableMatchesCluster                   1000  avgt   50    85200.096 ±   668.200  ns/op
UserPrivilegesBenchmark.measureTableMatchesCluster                  10000  avgt   50  1504118.062 ±  9832.737  ns/op
UserPrivilegesBenchmark.measureTableMatchesSchema                     100  avgt   50     4359.790 ±    51.477  ns/op
UserPrivilegesBenchmark.measureTableMatchesSchema                    1000  avgt   50    94705.489 ±  2712.242  ns/op
UserPrivilegesBenchmark.measureTableMatchesSchema                   10000  avgt   50  1624120.655 ±  9126.772  ns/op
```

### This branch
```
Benchmark                                                (privilegesSize)  Mode  Cnt    Score    Error  Units
UserPrivilegesBenchmark.measureTableMatches                           100  avgt   50   41.248 ±  1.786  ns/op
UserPrivilegesBenchmark.measureTableMatches                          1000  avgt   50   43.745 ±  1.851  ns/op
UserPrivilegesBenchmark.measureTableMatches                         10000  avgt   50   42.136 ±  0.516  ns/op
UserPrivilegesBenchmark.measureTableMatchesAnyOfCluster               100  avgt   50   42.485 ±  0.350  ns/op
UserPrivilegesBenchmark.measureTableMatchesAnyOfCluster              1000  avgt   50   42.213 ±  0.225  ns/op
UserPrivilegesBenchmark.measureTableMatchesAnyOfCluster             10000  avgt   50   43.278 ±  0.725  ns/op
UserPrivilegesBenchmark.measureTableMatchesAnyOfSchema                100  avgt   50   60.842 ±  0.411  ns/op
UserPrivilegesBenchmark.measureTableMatchesAnyOfSchema               1000  avgt   50   61.467 ±  0.408  ns/op
UserPrivilegesBenchmark.measureTableMatchesAnyOfSchema              10000  avgt   50   65.786 ±  0.531  ns/op
UserPrivilegesBenchmark.measureTableMatchesAnyOfTable                 100  avgt   50   24.765 ±  0.225  ns/op
UserPrivilegesBenchmark.measureTableMatchesAnyOfTable                1000  avgt   50   28.806 ±  0.446  ns/op
UserPrivilegesBenchmark.measureTableMatchesAnyOfTable               10000  avgt   50   31.643 ±  0.701  ns/op
UserPrivilegesBenchmark.measureTableMatchesCluster                    100  avgt   50  159.070 ±  3.365  ns/op
UserPrivilegesBenchmark.measureTableMatchesCluster                   1000  avgt   50  205.063 ± 21.505  ns/op
UserPrivilegesBenchmark.measureTableMatchesCluster                  10000  avgt   50  177.506 ± 11.164  ns/op
UserPrivilegesBenchmark.measureTableMatchesSchema                     100  avgt   50  130.063 ±  1.206  ns/op
UserPrivilegesBenchmark.measureTableMatchesSchema                    1000  avgt   50  161.101 ± 13.847  ns/op
UserPrivilegesBenchmark.measureTableMatchesSchema                   10000  avgt   50  132.127 ±  1.377  ns/op
```

Added the benchmark code just for reference, think it shouldn't be merged.

### CODE REVIEW

File: sql/src/main/java/io/crate/analyze/user/Privilege.java [hunk 1]
```diff
@@ -62,70 +60,24 @@
...
-        if (foundPrivilege == null) {
-            return false;
-        }
-        switch (foundPrivilege.state()) {
-            case GRANT:
-                return true;
-            case DENY:
-            default:
-                return false;
+    public static String schemaNameOfTableIdent(String ident) {
```
Comment (dobe) [inline sql/src/main/java/io/crate/analyze/user/Privilege.java]: what about moving this to the Schemas class as a static method called schemaNameFromIndexName. i think for example https://github.com/crate/crate/blob/7c1133685b294dfd5108bd845242409320e31e9c/sql/src/main/java/io/crate/operation/reference/sys/shard/ShardSchemaNameExpression.java#L36 would also benefit from it

File: sql/src/main/java/io/crate/operation/user/UserPrivileges.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+import java.util.Map;
+
+class UserPrivileges implements Iterable<Privilege> {
+
+    @Nullable
+    private static Privilege matchPrivilege(IntObjectHashMap<Privilege> privileges,
+                                            Privilege.Type type,
+                                            Privilege.Clazz clazz,
+                                            @Nullable String ident) {
+        Privilege foundPrivilege = privileges.get(Privilege.hashCode(Privilege.State.GRANT, type, clazz, ident));
```
Comment (dobe) [inline sql/src/main/java/io/crate/operation/user/UserPrivileges.java]: it is not safe to use a hashcode, those hashcodes can conflict. 
i think in order to optimize this you probably need to split out PrivilegeIdent from Privilege where PrivilegeIdent just has type, clazz and ident and use this as the key.

File: enterprise/users/src/main/java/io/crate/metadata/UsersPrivilegesMetaData.java [hunk 3]
```diff
@@ -65,16 +66,17 @@ public static UsersPrivilegesMetaData copyOf(@Nullable UsersPrivilegesMetaData o
         for (Map.Entry<String, Set<Privilege>> entry : oldMetaData.usersPrivileges.entrySet()) {
             Set<Privilege> privileges = new HashSet<>(entry.getValue().size());
             for (Privilege privilege : entry.getValue()) {
-                if (privilege.clazz().equals(Privilege.Clazz.TABLE) == false) {
-                    privileges.add(new Privilege(privilege.state(), privilege.type(), privilege.clazz(),
-                        privilege.ident(), privilege.grantor()));
+                PrivilegeIdent privilegeIdent = privilege.ident();
+                if (privilegeIdent.clazz().equals(Privilege.Clazz.TABLE) == false) {
+                    privileges.add(new Privilege(privilege.state(), privilegeIdent.type(), privilegeIdent.clazz(),
```
Comment (dobe) [inline enterprise/users/src/main/java/io/crate/metadata/UsersPrivilegesMetaData.java]: pls create a constructor with the PrivilegeIdent and reuse privilegeIdent, however i am not sure if we need targetIdent here?

Comment (seut) [inline enterprise/users/src/main/java/io/crate/metadata/UsersPrivilegesMetaData.java]: hm yes thats what it all is about, change the ident.

**Pull request merged**